File: day3a/rust/src/main.rs
use std::collections::HashSet;
use std::fs;

const FILE_PATH: &str = "../input.txt";

fn main() {
    let file_content = match fs::read_to_string(FILE_PATH) {
        Ok(str) => str,
        Err(err) => panic!("{err}"),
    };

    let lines = file_content.lines().filter(|line| !line.is_empty());

    let mut priority_sum = 0;
    for line in lines {
        let line_len = line.len();
        let half_length = line_len / 2;
        let first_half = &line[..half_length];
        let second_half = &line[half_length..line_len];

        let seen_chars: HashSet<char> = HashSet::from_iter(first_half.chars());
        for char in second_half.chars() {
            if seen_chars.contains(&char) {
                if char.is_uppercase() {
                    priority_sum += ((char as u8) - ('A' as u8) + 27) as i32;
                } else {
                    priority_sum += ((char as u8) - ('a' as u8) + 1) as i32;
                }
                break;
            }
        }
    }

    println!("{priority_sum}")
}

File: day3a/c++/main.cpp
#include <iostream>
#include <fstream>
#include <set>

const auto FILE_NAME = "../input.txt";

int main()
{
    std::ifstream input_file(FILE_NAME);

    if (input_file.is_open())
    {
        auto priority_sum = 0;

        std::string line;
        while (std::getline(input_file, line))
        {
            if (!line.empty())
            {

                auto line_len = line.length();
                auto half_len = line_len / 2;
                auto first_half = line.substr(0, half_len);
                auto second_half = line.substr(half_len);

                std::set<char> seen_chars;
                for (auto &ch : first_half)
                {
                    seen_chars.insert(ch);
                }

                // Traverse the string
                for (auto &ch : second_half)
                {
                    if (seen_chars.contains(ch))
                    {
                        // is uppercase
                        if (ch >= 'a')
                        {
                            priority_sum += ch - 'a' + 1;
                        }
                        else
                        {
                            priority_sum += ch - 'A' + 27;
                        }
                        break;
                    }
                }
            }
        }
        input_file.close();

        printf("%d\n", priority_sum);
    }
}
